app/transformers/PAYETransformer.scala

```diff
@@ -0,0 +1,300 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package transformers
+
+import models.Amount
+import play.api.Mode.Mode
+import play.api.{Configuration, Play}
+import play.api.libs.json._
+import play.api.libs.json.JsPath
+import uk.gov.hmrc.play.config.ServicesConfig
+
+trait PAYETransformer {
+  def middleTierAmountJson(name: String, value: Double, currency: String = "GBP"): JsObject = Json.obj(
+    name -> Json.obj(
+      "amount"   -> value,
+      "currency" -> currency
+    )
+  )
+
+  def middleTierRateJson(name: String, percent: Double): JsObject = Json.obj(
+    name -> Json.obj(
+      "percent" -> JsString(percent.toString + "%")
+    )
+  )
+
+  def appendAttribute(path: JsPath, anObject: JsObject) =
+    path.json.update((__).read[JsObject].map { o =>
+      o ++ anObject
+    })
+  def pickAmount(path: JsPath, json: JsValue): Option[Double] =
+    json.transform(path.json.pick[JsNumber]).asOpt.map(n => n.as[Double])
+}
+
+object PAYETransformer extends ServicesConfig {
+  protected def mode: Mode = Play.current.mode
+
+  protected def runModeConfiguration: Configuration = Play.current.configuration
+
+  val serviceUrl = baseUrl("tax-summaries-hod")
+
+  def middleTierJson(nino: String, year: Int): JsObject =
+    Json.obj( // add nino
+      "taxYear" -> JsNumber(year),
+      "nino"    -> JsString(nino),
+      "lnks" -> Json.arr(
+        Json.obj(
+          "rel"  -> "self",
+          "href" -> s"$serviceUrl/individuals/annual-tax-summary/$nino/$year"
+        )),
+      "allowance_data"     -> Json.obj("payload" -> Json.obj()),
+      "capital_gains_data" -> Json.obj("payload" -> Json.obj()),
+      "income_data"        -> Json.obj("payload" -> Json.obj()),
+      "income_tax"         -> Json.obj("payload" -> Json.obj(), "rates" -> Json.obj()),
+      "summary_data"       -> Json.obj("payload" -> Json.obj(), "rates" -> Json.obj()),
+      "gov_spending"       -> Json.obj()
+    )
+
+  implicit class RichJsObject(original: JsObject) extends PAYETransformer {
+    def omitEmpty: JsObject = original.value.foldLeft(original) {
+      case (obj, (key, JsString(st))) if st.isEmpty                                                   => obj - key
+      case (obj, (key, JsArray(arr))) if arr.isEmpty                                                  => obj - key
+      case (obj, (key, JsObject(child))) if new RichJsObject(JsObject(child)).omitEmpty.value.isEmpty => obj - key
+      case (obj, (_, _))                                                                              => obj
+    }
+
+    def safeTransform(jsonTransformer: Reads[JsObject]): JsObject =
```

Comment:
You can just return the Option and handle the None cases further up in the code. Making the decision here has meant that you need to call safeTransform in all public functions and that the error cannot be handled further up on the edge of your code